Implement new structure

Vannnu opened the pull request:
**What kind of change does this PR introduce?**
Feature

**What is the new behavior?**
New Model Structure and more...

**Does this PR introduce a breaking change?**
No

**Other information**:
N/A

Commit:
migrations: add strict_type and fix code_style

Commit:
feat(Controller: add two constants)

Commit:
feat(Controller) add methods to handle the services

Commit:
feat(Routes): add prefix and new endpoint

Commit:
feat(Request): add request for ingredient model

Commit:
feat(models): add a directory to contain models file and new struc of ingredient

Vannnu committed:
Merge branch 'develop' into feature/newStructure

Vannnu commented:
@ptesser Chiudo l'altra PR ho ricominciato il lavoro da qua. Ho incominciato intanto con gli ingredienti ho gia implementato il Crud e ho cominciato a spostare i modelli nella directory apposita.

ptesser commented on back-end/app/Http/Controllers/IngredientsController.php:
```diff
@@ -1,30 +1,68 @@
     {
-        return response()->json(['code' =>"Tutti gli ingredienti"],200);
+        $data = Ingredient::getAllIngredient();
+        return response()->json(['data' => $data ],self::OK);
```
spazio prima di `self`

Vannnu commented:
@ptesser mancano sempre gli unit test.

ptesser commented:
> @ptesser mancano sempre gli unit test.

Ok, apriti un task se intanto vuoi che mergiamo questa. In quella iniziale dei test sarebbe bello se impostassimo Travis CI per far runnare anche quelli.

Nel caso dai un occhio alle mie pull request così le mergio se sono ok.

Commit:
fix(Controller): add missing space

ptesser commented on back-end/app/Http/Controllers/IngredientsController.php:
```diff
@@ -1,30 +1,68 @@
-        return response()->json(['code'=>"Restituzione ingrediente"],200);
+        $name = $request->input('name');
+        $newIngredient = Ingredient::createIngredients($name);
+        return response()->json(['data' => $newIngredient],self::OK);
```
Stessa cosa qua e immagino anche sotto. Impostati PHPCS fixer nel caso :)

Vannnu replied:
mi pare di averli aggiunti. Si il fixer lo metto dammi tempo. :D No gli unit test li mettiamo in qua. Voglio controllare come inserire un DB di test.

ptesser replied:
Ok, allora metti sul titolo WIP, ho aggiunto un tool che se lo scrivi rende questa pull request not mergeable.

Commit:
:q
Merge branch 'feature/newStructure' of https://github.com/Hopstok/rock-recepies into feature/newStructure

ptesser commented:
@LucaVannucci gli ultimi due commit sono da sistemare come logica.

In una pull request non vogliamo avere commit che sistemano segnalazione della review. Perché non apportano nessun vantaggio alla storia del repository.

E' consigliato squashare quei commit in quelli dove avevi introdotto quel pezzo di codice.

ptesser commented on back-end/app/Http/Controllers/IngredientsController.php:
```diff
@@ -1,30 +1,68 @@
+    {
+        $data = Ingredient::getSpecifiedIngredients($id);
+        return response()->json(['data' => $data ], self::OK);
+
```
Spazi di troppo

Vannnu commented:
@ptesser  ma non li mettiamo nel changelog tanto è quello che conta. Non essere troppo nazi :d

ptesser commented:
> @ptesser ma non li mettiamo nel changelog tanto è quello che conta. Non essere troppo nazi :d

A parte che il changelog andrebbe generato dalle commit, quindi se ci sono cose sporche anche quello lo è :D comunque è una cosa semplice da fare.

Guardati https://fle.github.io/git-tip-keep-your-branch-clean-with-fixup-and-autosquash.html

Commit:
fix(Controller): change return on delete method

ptesser commented on back-end/app/Http/Controllers/IngredientsController.php:
```diff
@@ -1,30 +1,68 @@
+     *
+     * @return JsonResponse
+     */
+    public function delete($id)
```
Aggiungere tipo parametro e tipo di risposta.

ptesser commented on back-end/app/Http/Controllers/IngredientsController.php:
```diff
@@ -1,30 +1,70 @@
-     * @return \Illuminate\Http\Response
+     * @return \Illuminate\Http\JsonResponse
+     */
+    public function show($id): JsonResponse
```
Aggiungere tipo variabile.

ptesser commented on back-end/app/Http/Controllers/IngredientsController.php:
```diff
@@ -1,30 +1,70 @@
+    public function delete($id)
     {
-        return response()->json(['code'=>"Restituzione ingrediente"],200);
+        $deleteIngredient = Ingredient::deleteIngredient($id);
```
La delete se fallisce non emette un eccezione?

Di solito è consigliato gestire gli errori tramite eccezioni invece che con check sul tipo di ritorno.

ptesser commented on back-end/app/Models/Ingredient.php:
```diff
@@ -0,0 +1,82 @@
+     *
+     * @return Model
+     */
+    public static function createIngredients($name)
```
Aggiungere tipo del parametro e di ritorno, anche sotto dove non viene fatto.

ptesser replied:
Comunque io gestire il tutto tramite dependency injection.

ptesser commented on back-end/app/Models/Ingredient.php:
```diff
@@ -0,0 +1,82 @@
+            ->get();
+
+        if (count($ingredient) === 0) {
+
```
Rimuovere spazio.

ptesser commented on back-end/app/Models/Ingredient.php:
```diff
@@ -0,0 +1,82 @@
+     */
+    public static function deleteIngredient($id): bool
+    {
+        $ret = false;
```
Gestirei il tutto con un eccezione.

Commit:
feat(Controller,Models): modified flow with repository pattern

Commit:
fix(migration): add unique costraint to ingredients entity

Vannnu commented:
@ptesser  danni un occhio ci sono delle cose che non so se mi piacciono nella gestione delle risposte di ritorno http. La cancellazione dei commit la facciamo assieme. Aspetto tue ah le api sono funzionanti

Commit:
feat(Procedures): add handling for crud and repository pattern

Commit:
feat(Recipe): add handling for crud and repository pattern

Commit:
feat(Users): add handling for crud and repository pattern

Commit:
feat(route): add new endpoints

Commit:
feat(Ingredient): modify handling for crud and repository pattern

Commit:
feat(Recipe): modify handling for crud and repository pattern

Commit:
feat(Users): modify handling for crud and repository pattern

Commit:
fix(procedure): add fillable field

Commit:
feat(CommonImpl): change data get method

ptesser commented:
@LucaVannucci Dopo guardo il codice, comunque ricordati di usare il metodo che ti ho insegnato di `!fixup`. Di fatto, dalla nome dei commit:

Questo `feat(Ingredient): modify handling for crud and repository pattern` e questo `feat(Ingredient): add handling for crud and repository pattern`, ma anche per altri, sembrano riguardare la stessa funzionalità, quindi gli unirei

ptesser commented:
@LucaVannucci Ah ok, non avevo visto il tuo messaggio per la cancellazione delle commit insieme. Perfetto :)

ptesser commented on back-end/database/migrations/2017_03_29_145536_create_ingredients_table.php:
```diff
@@ -18,9 +21,8 @@ public function up()
-
             $table->increments('id');
-            $table->string('name',40);
+            $table->string('name',40)->unique();
```
Spazio `'name, 40'`

ptesser commented on back-end/database/migrations/2017_03_29_145508_create_recipes_table.php:
```diff
@@ -14,19 +17,15 @@ class CreateRecipesTable extends Migration
-
             $table->increments('id');
             $table->integer('user_id')->unsigned();
             $table->string('name',40);
```
Spazio anche qua, vedi sotto.

ptesser commented on back-end/app/Repositories/Interfaces/UserInt.php:
```diff
@@ -0,0 +1,17 @@
+ *
+ * interface to extend common methods.
+ */
+interface UserInt extends Common
```
Userei come nome solamente `User` e lo distinguerei nel suo utilizzo al massimo utilizzando il namespace. Stessa cosa anche per le altre interfacce.

A chi le usa deve risultare trasparente il fatto che tu stia usando un'interfaccia o meno. Come viene implementata la cosa non deve influenzare chi la utilizza.

ptesser commented on back-end/app/Repositories/Interfaces/UserInt.php:
```diff
@@ -0,0 +1,17 @@
+{
+    public function login();
+
+    public function passwordrecover();
```
Userei nomi del tipo `passwordRecover` e `passwordReset`

Vannnu replied:
Questa lasciala perdere che devo integrare Passport

ptesser commented on back-end/app/Repositories/Interfaces/Common.php:
```diff
@@ -0,0 +1,59 @@
+     *
+     * @return Collection
+     */
+     public function getById($id): array;
```
Aggiungere il tipo di `$id`

ptesser commented on back-end/app/Repositories/Interfaces/Common.php:
```diff
@@ -0,0 +1,59 @@
+     *
+     * @return mixed
+     */
+     public function update($id, array $attributes);
```
Aggiungere tipo su `$id`.

Credo che volendo possiamo anche tipare il tipo mixed. Vedi https://wiki.php.net/rfc/mixed-typehint

Però eviterei di tornare cose diverse. Penso sia meglio tornare sempre un certo tipo di valore e nel caso lanciare un'eccezione se c'è un errore.

ptesser commented on back-end/app/Repositories/Interfaces/Common.php:
```diff
@@ -0,0 +1,59 @@
+     *
+     * @return bool
+     */
+     public function delete($id): bool;
```
Aggiungere tipo

ptesser reviewed (commented):
@LucaVannucci Devo fare ancora, però in generale rivedrei in un modo più "sano" i vari nomi delle classi e interfacce.

La linea generale direi che è ok, però cercherei di spezzare il lavoro su più branch separate, altrimenti la review diventa troppo difficile da fare.

ptesser commented on back-end/app/Http/Controllers/Controller.php:
```diff
@@ -10,4 +10,25 @@
+     *
+     * @var int
+     */
+    const NOT_CREATED = 400;
```
Questo è Bad Request

ptesser commented on back-end/app/Repositories/Implementations/UserImpl.php:
```diff
@@ -0,0 +1,34 @@
+
+    }
+
+    public function passwordrecover()
```
Mi pareva di averlo già scritto da qualche parte, ma userei nome in camelCase, anche sotto sul reset.

Vannnu replied:
Lascia perdere le cose su user

Commit:
Wip

ptesser commented:
Chiusa perché troppo estesa. Si è deciso di aprirne di diverse più piccole.
